### ISSUE TITLE
COCOEval.summarize() mAP and AR calculations do not take into account class imbalances.

### ISSUE THREAD

**Issue by starlipernl** [pdollar/coco#382]
The summarize function of COCOEval does not take into account class imbalances. In other words, it takes the mean of precision and recalls across categories rather than taking a weighted average based on the numbers of objects in each category. This is giving skewed results when using the API for custom datasets.

**Comment by glenn-jocher** [comment]
@starlipernl yes this is the definition of mAP, the mean AP across all classes.

So class imbalances are irrelevant to the metric.

**Comment by rebeen** [comment]
Hi,
did you find how to get the score ?? I am doing object detection and with a pre-trained model but don't know how to get the following format

```
[{
"image_id": int, "category_id": int, "bbox": [x,y,width,height], "score": float,
}]
```

**Comment by glenn-jocher** [comment]
@rebeen see https://github.com/ultralytics/yolov3#map

**Comment by rebeen** [comment]
@glenn-jocher actually am  trying to learn object detection on COCO custom dataset so I am working on a small example with only 5 images so  I am using a pre-trained model I really appreciate if you can help  to do the evaluation I can share the code with you with my data if you have sometimes 

Thank you but I cannot understand your  script





# Training
```
for epoch in range(config.num_epochs):
    print(f"Epoch: {epoch}/{config.num_epochs}")
    model.train()
    i = 0
    for imgs, annotations in data_loader:
        i += 1
        imgs = list(img.to(device) for img in imgs)
        annotations = [{k: v.to(device) for k, v in t.items()} for t in annotations]
        loss_dict = model(imgs, annotations)
        losses = sum(loss for loss in loss_dict.values())

        optimizer.zero_grad()
        losses.backward()
        optimizer.step()

        print(f"Iteration: {i}/{len_dataloader}, Loss: {losses}")

```

**Comment by glenn-jocher** [comment]
@rebeen see https://github.com/ultralytics/yolov3/wiki/Train-Custom-Data for custom data training and testing.

**Comment by rebeen** [comment]
@glenn-jocher  Thank you very much but actually I don't want to change the format of coco I want to learn directly from coco as it is 
here is my results for coco based on YOLO format 
https://github.com/qfgaohao/pytorch-ssd/issues/105

which I follow this repo
https://github.com/mive93/pytorch-ssd

but I actually now have only 5 images and I want to learn how it works step by step 

I am really broken with SSD and object detection it is really really really harddddd

**Comment by glenn-jocher** [comment]
@rebeen https://github.com/ultralytics/yolov3/wiki/Train-Custom-Data works with any dataset including coco. Start there.

**Comment by rebeen** [comment]
@glenn-jocher  Thank you

**Comment by Michael-J98** [comment]
Hi~there.
Do you know how to calculate AR in coco metrics?
I found several definitions of AR which made me really confused.
And what does the "[maxium recall](http://cocodataset.org/#detection-eval)" refer to?
Thank you very much!

**Comment by timatom** [comment]
I feel like nobody adequately responded to @starlipernl's concern.

You are correct that the APs can be "skewed" relative to the final mAP from class imbalances. I also believe there are significantly more occurrences of humans in the COCO dataset, which actually can reduce the usefulness of the mAP. I would also add that this occurs when one class (or more), has APs that mostly perform significantly better than others.

Assuming the dataset's classes are balanced, the mAP is more useful than in the imbalanced case. How useful is debatable.

For more granular detail of your APs, I recommend evaluating additional statistics on your own if needed, e.g., pr-curve, confusion matrix, etc. Which statistics matter depends greatly on your use-case.

For those looking for a basic understand of mAP, I think the following post is nice: https://jonathan-hui.medium.com/map-mean-average-precision-for-object-detection-45c121a31173

Also, others should check out this repository for evaluating COCO formatted datasets: https://github.com/voxel51/fiftyone
It's a very good repository (not affiliated). It even does the weighted average calculations @starlipernl was suggesting.
